src/main/java/org/javarosa/xpath/expr/XPathEqExpr.java

```diff
@@ -70,7 +70,11 @@ public void writeExternal(DataOutputStream out) throws IOException {
     public static boolean testEquality(Object aval, Object bval) {
         boolean eq = false;
 
-        if (aval instanceof Boolean || bval instanceof Boolean) {
+        if (aval == null && bval == null) {
```

Comment:
Hi Shubham

Unfortunately this may require a deeper change.  The engine can't actually handle null values for XPath expression outputs, and it should be treated as an error for any values to ever be null.

I can't quite find a concrete code reference that describes this, although documentation in various places does reflect it:
https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/xpath/expr/FunctionUtils.java#L170

https://github.com/dimagi/commcare-core/blob/d9066ea561f4ce85717b25263f9ddad695c98b29/src/test/java/org/javarosa/core/model/data/test/StringDataTests.java#L50

Reply:
I see, I am running into this limitation while evaluating the dependent itemset choices for ex - `<itemset nodeset="instance('district')/district_list/district[state_id = $state]">` . Since this PR changes logic to initialize the fields with null values instead of empty string, the evalualtion for `[state_id = null]` fails due to a NPE in this function. It seems likely that we can still set a `null` variable as an empty string just for the purpose of evaluating itemsets without introducing any regressions. Does that sound like a good workaround on this ?

Reply:
I think the core question is whether you want the engine to be able to differentiate between a non-existent input and a blank input via the XPath Variable itself. It sounds like you don't need to, and that the null value is relevant for the URL endpoint generation but not the expressions here. I think the cleanest solution would definitely be that the Query Screen *always* populates a value for all of its inputs into the XPath evaluation context, and it checks for null inputs and replaces them with empty strings. That would be consistent with what the normal engine would do with a reference which is valid (IE: The node exists in the form structure) but is currently irrelevant or empty.